File: 7M4W/ehowlqk/boj_1074.js
const fs = require("fs");
const [N, r, c] = fs
  .readFileSync("./7M4W/ehowlqk/boj_1074.txt")
  .toString()
  .split(" ")
  .map((e) => Number(e));

function z(size, row, col) {
  const loc = zone(2 ** size, row, col);
  if (size === 1) {
    return loc - 1;
  }
  let order = 0;
  const div = 2 ** (size - 1);
  const add = (2 ** size) ** 2 / 4;
  switch (loc) {
    case 4:
      order += add;
    case 3:
      order += add;
    case 2:
      order += add;
    default:
      break;
  }

  return order + z(size - 1, row % div, col % div);
}

function zone(size, row, col) {
  if (row < size / 2) {
    if (col < size / 2) {
      return 1;
    } else return 2;
  } else {
    if (col < size / 2) {
      return 3;
    } else return 4;
  }
}

console.log(z(N, r, c));

File: 7M4W/minjeeki/boj_1074.py
N, r, c = map(int, input().split())

def find_num(N, r, c):
    if N == 0:
        return 0
    
    # 4등분 할 경우 한 변의 길이
    size = 2 ** (N - 1)
    # 현재 좌표의 위치 파악
    row, col = r // size, c // size
    
    num = 0
    # 왼쪽 위
    if row == 0 and col == 0:
        num = find_num(N - 1, r, c)
    # 오른쪽 위
    elif row == 0 and col == 1:
        num = size ** 2 + find_num(N - 1, r, c - size)
    # 왼쪽 아래
    elif row == 1 and col == 0:
        num = 2 * size ** 2 + find_num(N - 1, r - size, c)
    # 오른쪽 아래
    else:
        num = 3 * size ** 2 + find_num(N - 1, r - size, c - size)
    
    return num

print(find_num(N, r, c))